tests/handlers/test_directory.py

```diff
@@ -432,6 +432,60 @@ def test_allowed(self):
         self.assertEquals(200, channel.code, channel.result)
 
 
+class TestCreatePublishedRoomACL(unittest.HomeserverTestCase):
+    user_id = "@admin:test"
+    denied_user_id = "@test:test"
+    data = {"room_alias_name": "unofficial_test"}
+
+    servlets = [directory.register_servlets, room.register_servlets]
+
+    def prepare(self, reactor, clock, hs):
+        # This time we add custom room list publication rules
+        config = {}
+        config["alias_creation_rules"] = []
+        config["room_list_publication_rules"] = [
+            {"user_id": "*", "alias": "*", "action": "deny"},
+            {"user_id": "@admin:test", "alias": "*", "action": "allow"},
+        ]
+
+        rd_config = RoomDirectoryConfig()
+        rd_config.read_config(config)
+
+        self.hs.config.roomdirectory.is_publishing_room_allowed = (
+            rd_config.is_publishing_room_allowed
+        )
+
+        return hs
+
+    def test_denied(self):
+        # NOTE Setting is_public=True isn't enough
+        self.data["visibility"] = "public"
+        self.helper.create_room_as(
+            self.denied_user_id, extra_content=self.data, expect_code=403
+        )
+
+    def test_allowed_without_publish(self):
+        self.helper.create_room_as(
+            self.denied_user_id,
+            extra_content=self.data,
+            is_public=False,
+            expect_code=200,
+        )
+
+    def test_allowed_as_allowed(self):
+        self.helper.create_room_as(
+            self.user_id, extra_content=self.data, is_public=False, expect_code=200
+        )
+
+    def test_denied_then_retry_without_publish(self):
+        self.test_denied()
+        self.test_allowed_without_publish()
```

Comment:
Comment here would be similarly helpful, something like this?

```suggestion
    def test_denied_then_retry_without_publish(self):
        # Try to create a public room with alias without permission. This should be rejected.
        self.test_denied()
        # Retry as the same user, but without publishing to the directory. This should be accepted.
        self.test_allowed_without_publish()
```